File: src/api/schemas/deck_suggestions.py
"""Schemas for the deck-suggestion request API (F172)."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from src.deck_suggestions.repository import colors_from_str, load_result

SUGGESTION_FORMATS = {
    "commander",
    "standard",
    "pioneer",
    "modern",
    "legacy",
    "vintage",
    "pauper",
    "casual",
}
SUGGESTION_ARCHETYPES = {"aggro", "control", "midrange", "combo", "tempo", "ramp"}
SUGGESTION_STATUSES = ("pending", "processing", "done", "failed")
MAX_NOTES_LENGTH = 1000


class DeckSuggestionCreate(BaseModel):
    format_name: str
    commander_card_id: int | None = None
    colors: list[str] = Field(default_factory=list)
    archetype: str | None = None
    notes: str | None = Field(default=None, max_length=MAX_NOTES_LENGTH)


class DeckSuggestionSave(BaseModel):
    deck_name: str | None = Field(None, min_length=1, max_length=300)


class SuggestionSummary(BaseModel):
    total_cards: int
    owned_cards: int
    missing_cards: int
    missing_cost_brl: float | None = None
    unresolved_count: int = 0


class DeckSuggestionSchema(BaseModel):
    id: int
    format_name: str
    commander_card_id: int | None = None
    commander_name: str | None = None
    colors: list[str] = Field(default_factory=list)
    archetype: str | None = None
    notes: str | None = None
    status: Literal["pending", "processing", "done", "failed"]
    error_message: str | None = None
    saved_deck_id: int | None = None
    created_at: datetime
    processed_at: datetime | None = None
    summary: SuggestionSummary | None = None
    result: dict | None = None


class DeckSuggestionSavedResponse(BaseModel):
    deck_id: int


def _summary_from_result(result: dict | None) -> SuggestionSummary | None:
    if not result or not isinstance(result.get("summary"), dict):
        return None
    try:
        return SuggestionSummary(**result["summary"])
    except (TypeError, ValueError):
        return None


def to_schema(row, include_result: bool) -> DeckSuggestionSchema:
    """Convert a ``DeckSuggestionRequestRow`` into its API schema.

    ``summary`` is filled for done requests; the full ``result`` only when
    ``include_result`` is True (detail endpoint).
    """
    result = load_result(row) if row.status == "done" else None
    return DeckSuggestionSchema(
        id=row.id,
        format_name=row.format_name,
        commander_card_id=row.commander_card_id,
        commander_name=row.commander_name,
        colors=colors_from_str(row.colors),
        archetype=row.archetype,
        notes=row.notes,
        status=row.status,
        error_message=row.error_message,
        saved_deck_id=row.saved_deck_id,
        created_at=row.created_at,
        processed_at=row.processed_at,
        summary=_summary_from_result(result),
        result=result if include_result else None,
    )

File: frontend/src/types/deckSuggestions.ts
/**
 * Deck suggestion types (F172). Mirrors the backend contract in
 * `src/api/schemas/deck_suggestions.py` — field names are snake_case as
 * returned by the API.
 */

export type SuggestionStatus = "pending" | "processing" | "done" | "failed";

export interface DeckSuggestionCreate {
  format_name: string;
  commander_card_id?: number | null;
  colors?: string[];
  archetype?: string | null;
  notes?: string | null;
}

export interface SuggestionSummary {
  total_cards: number;
  owned_cards: number;
  missing_cards: number;
  missing_cost_brl: number | null;
  unresolved_count: number;
}

export interface SuggestedCard {
  name_en: string;
  quantity: number;
  category: string;
  reason: string | null;
  card_id: number | null;
  set_code: string | null;
  collector_number: string | null;
  image_uri: string | null;
  is_owned: boolean;
  owned_quantity: number;
  missing_quantity: number;
  unit_price: number | null;
  missing_cost: number | null;
}

export interface SuggestionResult {
  deck_name: string;
  strategy: string;
  format_name: string;
  commander: { name_en: string; card_id: number | null } | null;
  cards: SuggestedCard[];
  summary: SuggestionSummary;
  unresolved: string[];
  warnings: string[];
  provider: string;
  model: string;
  generated_at: string;
}

export interface DeckSuggestion {
  id: number;
  format_name: string;
  commander_card_id: number | null;
  commander_name: string | null;
  colors: string[];
  archetype: string | null;
  notes: string | null;
  status: SuggestionStatus;
  error_message: string | null;
  saved_deck_id: number | null;
  created_at: string;
  processed_at: string | null;
  summary: SuggestionSummary | null;
  result?: SuggestionResult | null;
}

export interface DeckSuggestionSaveResult {
  deck_id: number;
}
